ERCS/erc-7811.md

```diff
@@ -0,0 +1,239 @@
+---
+eip: 7811
+title: Wallet Asset Discovery
+description: JSON-RPC method for wallets to share a user’s full asset list with Dapps, including assets not easily found through on-chain data alone
+author: Luka Isailovic (@lukaisailovic), Konrad Kopp (@kopy-kat), Derek Rein (@arein)
+discussions-to: https://ethereum-magicians.org/t/erc-7811-wallet-asset-discovery/21639
+status: Draft
+type: Standards Track
+category: ERC
+created: 2024-11-07
+requires: 20, 155, 721, 5792
+---
+
+## Abstract
+
+This ERC introduces a new RPC call, `wallet_getAssets`, for wallets to declare to the Dapp what assets are owned by the user. This allows for more accurate asset discovery and the use of assets that aren’t available on-chain but can be provided by the wallet
+
+## Motivation
+
+Currently, Dapps primarily rely on on-chain data to determine a user's balance, which can be limiting. Furthermore, a Dapp might restrict the user from initiating actions that the wallet could otherwise resolve, as it cannot account for the total assets a user has across different accounts or chains.
+
+Wallets already have information about a user's assets, including those not visible on-chain, and need a way to communicate that information to Dapps.
+
+## Specification
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+### Method: `wallet_getAssets`
+
+#### Request schema
+
+```ts
+type Hex = `0x${string}`;
+
+type WalletGetAssetsRequest = {
+  account: Hex;
+  assetFilter?: Record<Hex, string[]>;
+  assetTypeFilter?: string[];
+  chainFilter?: Hex[];
+};
+```
+
+`account` is a **REQUIRED** field that indicates for which account assets are requested.
+
+If the `assetFilter` field is provided, the wallet **MUST** only return the assets specified within it, effectively disregarding the `assetTypeFilter` and `chainFilter` fields as they are already implicitly defined within the `assetFilter`.
+
+If the `assetFilter` field is omitted, the wallet **SHOULD** return all available assets for the requested account. It is **RECOMMENDED** that the returned assets be ordered by estimated value in descending order, as determined by the wallet.
+
+`assetTypeFilter` is an **OPTIONAL** field that specifies an array of asset types, as defined in this ERC. If `assetTypeFilter` field is provided, wallet **MUST** include only those assets in the response.
+
+`chainFilter` is an **OPTIONAL** field that specifies an array of chain ids, where each value in the array **MUST** be a valid [EIP-155](./eip-155.md) chainId
+
+#### Example request
+
+```json
+{
+  "account": "0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045",
+  "assetFilter": {
+    "0x1": ["0x456", "native"],
+    "0xa": ["0x789", "native"]
+  },
+  "assetTypeFilter": ["ERC20", "native"],
+  "chainFilter": ["0x1"]
+}
+```
+
+#### Response schema
+
+```ts
+type Asset = {
+  address: Hex | "native";
```

Comment:
`address` is not sufficient to identify the asset. For example ERC-720 and ERC-1155 tokens have an additional `tokenId`. This problem is also addressed in [CAIP-19](https://github.com/ChainAgnostic/namespaces/blob/main/eip155/caip19.md)

Furthermore, `address` doesn't make sense for native tokens. We handled that with a special case, but I'd argue `address` should be moved into `metadata` instead.